controllers/etcd_controller.go

```diff
@@ -414,59 +357,90 @@ func (r *EtcdReconciler) delete(ctx context.Context, etcd *druidv1alpha1.Etcd) (
-	pdb := &policyv1beta1.PodDisruptionBudget{}
-	err := r.Get(ctx, types.NamespacedName{Name: etcd.Name, Namespace: etcd.Namespace}, pdb)
-
-	if err == nil {
-		// pdb already exists, claim it
-
-		selector, err := metav1.LabelSelectorAsSelector(etcd.Spec.Selector)
-		if err != nil {
-			logger.Error(err, "Error converting etcd selector to selector")
-			return err
-		}
 
-		logger.Info("Claiming pdb object")
-		_, err = r.claimPodDisruptionBudget(ctx, etcd, selector, pdb)
+	// Find out k8s version
+	targetCoreClient, err := k8s.NewForConfig(r.Config)
+	if err != nil {
 		return err
 	}
-
-	if !apierrors.IsNotFound(err) {
+	targetServerVersion, err := targetCoreClient.Discovery().ServerVersion()
+	if err != nil {
 		return err
 	}
-
-	// Required podDisruptionBudget doesn't exist. Create new
-	pdb, err = r.getPodDisruptionBudgetFromEtcd(etcd, values, logger)
+	targetKubernetesVersion, err := semver.NewVersion(targetServerVersion.GitVersion)
 	if err != nil {
 		return err
 	}
 
-	logger.Info("Creating PodDisruptionBudget", "poddisruptionbudget", kutil.Key(pdb.Namespace, pdb.Name).String())
-	if err := r.Create(ctx, pdb); err != nil {
-		return err
-	}
+	if k8sutils.ConstraintK8sGreaterEqual121.Check(targetKubernetesVersion) {
+		pdb := &policyv1.PodDisruptionBudget{}
 
-	return nil
-}
+		err := r.Get(ctx, types.NamespacedName{Name: etcd.Name, Namespace: etcd.Namespace}, pdb)
+		if err == nil {
+			// pdb already exists
+			return nil
+		}
+		if !apierrors.IsNotFound(err) {
+			return err
+		}
 
-func (r *EtcdReconciler) getPodDisruptionBudgetFromEtcd(etcd *druidv1alpha1.Etcd, values map[string]interface{}, logger logr.Logger) (*policyv1beta1.PodDisruptionBudget, error) {
-	var err error
-	decoded := &policyv1beta1.PodDisruptionBudget{}
-	pdbPath := getChartPathForPodDisruptionBudget()
-	chartPath := getChartPath()
-	renderedChart, err := r.chartApplier.Render(chartPath, etcd.Name, etcd.Namespace, values)
-	if err != nil {
-		return nil, err
-	}
-	if content, ok := renderedChart.Files()[pdbPath]; ok {
-		decoder := yaml.NewYAMLOrJSONDecoder(bytes.NewReader([]byte(content)), 1024)
-		if err = decoder.Decode(&decoded); err != nil {
-			return nil, err
+		// Required podDisruptionBudget doesn't exist. Create new
+		pdbPath := getChartPathForPodDisruptionBudget()
+		chartPath := getChartPath()
+		renderedChart, err := r.chartApplier.Render(chartPath, etcd.Name, etcd.Namespace, values)
+		if err != nil {
+			return err
+		}
+		if content, ok := renderedChart.Files()[pdbPath]; ok {
+			decoder := yaml.NewYAMLOrJSONDecoder(bytes.NewReader([]byte(content)), 1024)
+			if err = decoder.Decode(&pdb); err != nil {
+				return err
+			}
+		}
+		if err != nil {
+			return err
+		}
+
+		logger.Info("Creating PodDisruptionBudget", "poddisruptionbudget", kutil.Key(pdb.Namespace, pdb.Name).String())
+		if err := r.Create(ctx, pdb); err != nil {
+			return err
+		}
+	} else {
+		pdb := &policyv1beta1.PodDisruptionBudget{}
```

Comment:
There is a lot of code repetition. One proposal is to use generics.  Something similar to the following code. I have not tested it obviously but this just hints at how to use it.
Unfortunately go generics is quite half baked (that is what happens when it is not included in the original language) so you cannot use generics for methods - https://github.com/golang/go/issues/49085
```go
func createPdbIfNotExists[T client.Object](ctx context.Context, cl client.Client, chartApplier kubernetes.ChartApplier, etcdNsName types.NamespacedName, values map[string]interface{}, pdbObj T) error {
	err := cl.Get(ctx, etcdNsName, pdbObj)
	if err == nil {
		return nil
	}
	if !apierrors.IsNotFound(err) {
		return err
	}

	pdbPath := getChartPathForPodDisruptionBudget()
	chartPath := getChartPath()
	renderedChart, err := chartApplier.Render(chartPath, etcdNsName.Name, etcdNsName.Namespace, values)
	if err != nil {
		return err
	}
	if content, ok := renderedChart.Files()[pdbPath]; ok {
		decoder := yaml.NewYAMLOrJSONDecoder(bytes.NewReader([]byte(content)), 1024)
		if err = decoder.Decode(&pdbObj); err != nil {
			return err
		}
	}
	if err != nil {
		return err
	}

	if err := cl.Create(ctx, pdbObj); err != nil {
		return err
	}
	return nil
}
```